### PR TITLE
4.x deprecated VideoCapture backends

### PR DESCRIPTION
**Pull request by vovka643**
Added list of deprecated backends, which not supported by opencv anymore. This needs to give for user more information about chosen backend, when VideoCapture failed trying to use chosen by user  backend. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/videoio/src/cap.cpp [hunk 1]
```diff
@@ -315,6 +316,19 @@ bool VideoCapture::open(int cameraNum, int apiPreference, const std::vector<int>
         CV_Error_(Error::StsError, ("could not open camera %d", cameraNum));
     }
 
+    for (size_t i = 0; i < deprecatedBackends.size(); i++)
+    {
+        const VideoBackendInfo& info = deprecatedBackends[i];
+        if (apiPreference == info.id)
+        {
+            //chosen backend is not supported by OpenCV
+            CV_LOG_INFO(NULL, "VIDEOIO: chosen backend is not supported by OpenCV");
```
Comment (asmorkalov) [inline modules/videoio/src/cap.cpp:325]: It's better to stress that the backend removed from OpenCV.

File: modules/videoio/src/videoio_registry.cpp [hunk 2]
```diff
@@ -183,6 +188,20 @@ static const struct VideoBackendInfo builtin_backends[] =
...
+#endif
+    DECLARE_DEPRECATED_BACKEND(CAP_QT, "QuickTime")
+    DECLARE_DEPRECATED_BACKEND(CAP_UNICAP, "Unicap")
+    DECLARE_DEPRECATED_BACKEND(CAP_OPENNI, "OpenNI")
+    DECLARE_DEPRECATED_BACKEND(CAP_OPENNI_ASUS, "OpenNI")
+    DECLARE_DEPRECATED_BACKEND(CAP_GIGANETIX, "GigEVisionSDK")
+};
+
+
+
```
Comment (asmorkalov) [inline modules/videoio/src/videoio_registry.cpp:204]: Redundant empty lines.

**Review by asmorkalov — APPROVED** [review]

File: modules/videoio/src/cap.cpp [hunk 3]
```diff
@@ -315,6 +316,19 @@ bool VideoCapture::open(int cameraNum, int apiPreference, const std::vector<int>
...
     }
 
+    for (size_t i = 0; i < deprecatedBackends.size(); i++)
+    {
+        const VideoBackendInfo& info = deprecatedBackends[i];
+        if (apiPreference == info.id)
+        {
+            //chosen backend is removed from OpenCV
+            CV_LOG_INFO(NULL, "VIDEOIO: chosen backend is removed from OpenCV");
+            return false;
```
Comment (alalek) [inline modules/videoio/src/cap.cpp:326]: This message is not informative (it doesn't show even `id` or the `name`).
Code comment is redundant here.

File: modules/videoio/src/videoio_registry.cpp [hunk 4]
```diff
@@ -183,6 +188,20 @@ static const struct VideoBackendInfo builtin_backends[] =
     // dropped backends: MIL, TYZX
 };
 
+static const struct VideoBackendInfo deprecated_backends[] =
+{
+#ifdef WIN32
```
Comment (alalek) [inline modules/videoio/src/videoio_registry.cpp:193]: > `WIN32`

Must be `_WIN32`

**Comment by asmorkalov** [comment]
@opencv-alalek Could you take a look on the final version?

File: modules/videoio/src/videoio_registry.cpp [hunk 5]
```diff
@@ -179,10 +184,23 @@ static const struct VideoBackendInfo builtin_backends[] =
 #ifdef HAVE_OBSENSOR
     DECLARE_STATIC_BACKEND(CAP_OBSENSOR, "OBSENSOR", MODE_CAPTURE_BY_INDEX, 0, create_obsensor_capture, 0)
 #endif
-
```
Comment (opencv-alalek) [inline modules/videoio/src/videoio_registry.cpp:182]: Unnecessary unrelated whitespace change.

File: modules/videoio/src/videoio_registry.cpp [hunk 6]
```diff
@@ -329,6 +352,18 @@ class VideoBackendRegistry
...
+
+    inline std::vector<VideoBackendShortInfo> getDeprecatedBackends() const
+    {
+        std::vector<VideoBackendShortInfo> result;
+        for (size_t i = 0; i < deprecatedBackends.size(); i++)
+        {
+            const VideoBackendShortInfo& info = deprecatedBackends[i];
+            result.push_back(info);
+        }
+        return result;
```
Comment (opencv-alalek) [inline modules/videoio/src/videoio_registry.cpp:364]: Why do we need manual `for` loop here?

File: modules/videoio/src/videoio_registry.cpp [hunk 7]
```diff
@@ -362,6 +404,15 @@ cv::String getBackendName(VideoCaptureAPIs api)
         if (backend.id == api)
             return backend.name;
     }
+
+    const int M = sizeof(deprecated_backends) / sizeof(deprecated_backends[0]);
+    for (size_t i = 0; i < M; i++)
+    {
+        const VideoBackendShortInfo& backend = deprecated_backends[i];
+        if (backend.id == api)
+            return cv::format("DerecatedVideoAPI(%d)", (int)api);
```
Comment (opencv-alalek) [inline modules/videoio/src/videoio_registry.cpp:413]: typo

File: modules/videoio/src/videoio_registry.cpp [hunk 8]
```diff
@@ -449,6 +500,17 @@ std::string getCameraBackendPluginVersion(VideoCaptureAPIs api,
...
+
+    const std::vector<VideoBackendShortInfo> derpecatedBackends = VideoBackendRegistry::getInstance().getDeprecatedBackends();
+    for (size_t i = 0; i < derpecatedBackends.size(); i++)
+    {
+        const VideoBackendShortInfo& info = derpecatedBackends[i];
+        if (api == info.id)
+        {
+            CV_Error(Error::StsError, cv::format("Backend ID %i is removed from OpenCV", api));
+        }
+    }
```
Comment (opencv-alalek) [inline modules/videoio/src/videoio_registry.cpp:512]: We don't need that change here.

It is out of scope of initial request.

File: modules/videoio/src/videoio_registry.cpp [hunk 9]
```diff
@@ -489,6 +562,17 @@ std::string getWriterBackendPluginVersion(VideoCaptureAPIs api,
...
```
Comment (opencv-alalek) [inline modules/videoio/src/videoio_registry.cpp:574]: We don't need that change here.

**Review by opencv-alalek — COMMENTED** [review]
@asmorkalov I expecting minimal code review from you as you are the primary mentor of that change.

File: modules/videoio/src/cap.cpp [hunk 10]
```diff
@@ -315,6 +333,22 @@ bool VideoCapture::open(int cameraNum, int apiPreference, const std::vector<int>
...
+            CV_LOG_DEBUG(NULL,
+                cv::format("VIDEOIO(%s): backend is removed from OpenCV",
+                    info.name));
+            return false;
+        }
+    }
+
+    CV_LOG_DEBUG(NULL, cv::format("VIDEOIO: choosen backend does not work or wrong."
+        "Please make sure that your computer support chosen backend and OpenCV built"
+        "with right flags."));
```
[the remark this answers is not in the document]
Comment (vovka643) [reply modules/videoio/src/cap.cpp:350]: For separating messages we need to create new list of all supported backends. 
available_backends creates when opencv builds. It depends on build flags and specifications of computer. We can't say is this api number supported or just an error without list of all supported backends.

This message will not show for `CAP_ANY`. I double checked it.

File: modules/videoio/src/cap.cpp [hunk 11]
```diff
@@ -618,6 +652,22 @@ bool VideoWriter::open(const String& filename, int apiPreference, int fourcc, do
...
     }
+
+    for (const auto& info : videoio_registry::getDeprecatedBackends())
+    {
+        if (apiPreference == info.id)
+        {
+            CV_WRITER_LOG_DEBUG(NULL,
+                cv::format("VIDEOIO(%s): backend is removed from OpenCV",
+                    info.name));
+            return false;
```
[the remark this answers is not in the document]
Comment (opencv-alalek) [reply modules/videoio/src/cap.cpp:663]: > can't

What is the problem?

https://github.com/opencv/opencv/commit/872bb3255a9d87ee4be784b84fd30226a34d911e


---

> We need name of deprecated backend

There are changes in `getBackendName()`. Probably they are incomplete.
Anyway they are not required for `checkDeprecatedBackend()`.

File: modules/videoio/src/cap.cpp [hunk 12]
```diff
@@ -117,6 +117,7 @@ bool VideoCapture::open(const String& filename, int apiPreference, const std::ve
 
     const VideoCaptureParameters parameters(params);
     const std::vector<VideoBackendInfo> backends = cv::videoio_registry::getAvailableBackends_CaptureByFilename();
+    const std::vector<VideoDeprecatedBackendInfo> deprecatedBackends = cv::videoio_registry::getDeprecatedBackends();
```
Comment (opencv-alalek) [inline modules/videoio/src/cap.cpp:120]: Not used anymore.

Function itself is not needed.

Structure should be dropped too.

File: modules/videoio/src/videoio_registry.cpp [hunk 13]
```diff
@@ -449,6 +500,7 @@ std::string getCameraBackendPluginVersion(VideoCaptureAPIs api,
             return getCapturePluginVersion(info.backendFactory, version_ABI, version_API);
         }
     }
+
```
Comment (opencv-alalek) [inline modules/videoio/src/videoio_registry.cpp:503]: unnecessary unrelated whitespace change here and below

**Review by opencv-alalek — COMMENTED** [review]
Squash commits into one to see unnecessary changes.

File: modules/videoio/src/videoio_registry.cpp [hunk 14]
```diff
@@ -175,14 +180,25 @@ static const struct VideoBackendInfo builtin_backends[] =
...
+static const struct VideoDeprecatedBackendInfo deprecated_backends[] =
+{
+#ifdef _WIN32
+    DECLARE_DEPRECATED_BACKEND(CAP_VFW, "Video for Windows")
+#endif
+    DECLARE_DEPRECATED_BACKEND(CAP_QT, "QuickTime")
+    DECLARE_DEPRECATED_BACKEND(CAP_UNICAP, "Unicap")
+    DECLARE_DEPRECATED_BACKEND(CAP_OPENNI, "OpenNI")
+    DECLARE_DEPRECATED_BACKEND(CAP_OPENNI_ASUS, "OpenNI")
+    DECLARE_DEPRECATED_BACKEND(CAP_GIGANETIX, "GigEVisionSDK")
```
Comment (opencv-alalek) [inline modules/videoio/src/videoio_registry.cpp:198]: We don't need macro here, see https://github.com/opencv/opencv/commit/872bb3255a9d87ee4be784b84fd30226a34d911e

**Review by asmorkalov — APPROVED** [review]
:+1:
